consumer_group.go

```diff
@@ -220,6 +312,269 @@ func (c *consumerGroup) Consume(ctx context.Context, topics []string, handler Co
+		// group metadata
+		if members.UserData != nil {
+			c.userData = members.UserData
+		} else {
+			c.userData = c.config.Consumer.Group.Member.UserData
+		}
+
+		for _, partitions := range claims {
+			sort.Sort(int32Slice(partitions))
+		}
+	}
+
+	newAssignedPartitions := diffAssignment(claims, c.ownedPartitions)
+	if c.protocol == COOPERATIVE {
+		revokedPartitions := diffAssignment(c.ownedPartitions, claims)
+		Logger.Printf("updating consumer(group:%s, member:%s, generation:%d, isLeader:%v)\n"+
+			"*** All Assignments: %v\n"+
+			"*** New Partitions: %v\n"+
+			"*** Revoked Partitions: %v\n",
+			c.groupID, c.memberID, c.generationID, c.isLeader,
+			claims, newAssignedPartitions, revokedPartitions)
+
+		if len(revokedPartitions) > 0 {
+			err = c.revokedPartitions(revokedPartitions)
+			if err != nil {
+				Logger.Printf("error when revoking partitions: %v", err)
+			}
+			cancelFunc()
+		}
+	}
+	c.ownedPartitions = claims
+
+	err = c.startNewPartitions(newAssignedPartitions)
+	if err != nil {
+		c.revokedOwnedPartitions()
+		cancelFunc()
+	}
+
+	// only the leader needs to check whether there are newly-added partitions in order to trigger a rebalance
+	if c.isLeader {
+		go c.loopCheckPartitionNumbers(ctx, cancelFunc, c.allSubscribedTopicPartitions, c.allSubscribedTopics)
+	}
+
+	select {
+	case <-c.closed:
+		cancelFunc()
+	case <-ctx.Done():
+	}
+	Logger.Printf("consumer(group:%s, member:%s, generation:%d, isLeader:%v) context is done\n",
+		c.groupID, c.memberID, c.generationID, c.isLeader)
+
+	// if using EAGER rebalance protocol, we need to revoke all owned partitions before sending new JoinGroupRequest
+	if c.protocol == EAGER {
+		c.revokedOwnedPartitions()
+	}
+
+	// make sure heartbeat loop is stopped
+	close(hbDying)
+	<-hbDone
+	return nil
+}
+
+func (c *consumerGroup) joinGroup(ctx context.Context, topics []string, retries int) (*SyncGroupResponse, error) {
+	coordinator, err := c.joinPrepare(topics)
+	if err != nil {
+		return c.retryJoinGroup(ctx, topics, retries-1, err)
+	}
+
+	// todo: add metrics
+
+	// Join consumer group
+	join, err := c.joinGroupRequest(coordinator, topics)
+	if err != nil {
+		_ = coordinator.Close()
+		return nil, err
+	}
+
+	switch join.Err {
+	case ErrNoError:
+		c.memberID = join.MemberId
+	case ErrUnknownMemberId, ErrIllegalGeneration:
+		// reset member ID and retry immediately
+		c.memberID = ""
+		return c.joinGroup(ctx, topics, retries)
+	case ErrNotCoordinatorForConsumer, ErrRebalanceInProgress, ErrOffsetsLoadInProgress:
+		// retry after backoff
+		return c.retryJoinGroup(ctx, topics, retries, join.Err)
+	case ErrMemberIdRequired:
+		// from JoinGroupRequest v4 onwards (due to KIP-394) if the client starts
+		// with an empty member id, it needs to get the assigned id from the
+		// response and send another join request with that id to actually join the
+		// group
+		c.memberID = join.MemberId
+		return c.retryJoinGroup(ctx, topics, retries+1 /*keep retry time*/, join.Err)
```

Comment:
Similar to the accepted pull request #2759, should this be `return c.joinGroup(ctx, topics, retries)`?